Kh agent/vibe dev/pwa instructions

bromiesTM opened the pull request.

bromiesTM committed:
c/feat(PWA): add pwa banner & install instructions dialog

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

thlehmann-ionos commented on src/lib/IONOS/components/common/Dialog.svelte:
```diff
@@ -11,12 +12,17 @@
 
 	$: if (show) {
 		el?.showModal();
+		closeButton?.showModal();
```
Correct problem, wrong reasoning.

thlehmann-ionos commented on src/lib/IONOS/components/common/Dialog.svelte:
```diff
@@ -41,6 +48,25 @@
 	}
 </script>
 
+{#if externalClose}
```
What did you try to actually solve here? How is this related to the PWA install notification?
Rendering another dialog within a dialog seems odd.
Let's discuss.

thlehmann-ionos replied:
From direct conversation: this is used to render a close X button **outside** (above) of the `<dialog>` element, which is not possible with `position`. 

Rendering another `<dialog>` is probably the only feasible solution. However this is not self-explanatory/speaking and should at least get code comments why this exists.

thlehmann-ionos replied:
1. Please commit this change as its own commit and explain why it was made.
2. I wonder whether moving the dialog to its own component with a proper name could help making this more self-explanatory (thereby resolving my initial confusion).

thlehmann-ionos replied:
It does not work in Chrome. First, due to the `showModal()` it overlays the associated dialog. With `show()` it would not be visible at all. 

I think we should change the inner layout of the `<Dialog>` to render that close button. The dialog outline would be a child element of `<dialog>` as is the "external" close .

bromiesTM committed:
c/chore: add styleguide test pages

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -3,17 +3,33 @@
 	const DAYS = 24 * 60 * 60 * 1000;
 
+	// PWA Installation state
+	let deferredPrompt: BeforeInstallPromptEvent | null = null;
```
This variable should have a speaking name and the comment should be removed.

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -50,16 +66,108 @@
+		}
+	};
+
+	const handlePWADismiss = () => {
```
This is only used on time. Call `dismissPWAPrompt()` directly where this function is called.

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -50,16 +66,108 @@
 	{/each}
 </div>
+
+<!-- PWA Installation Dialog -->
```
This comment add no additional value.

thlehmann-ionos commented on src/lib/IONOS/services/pwa.ts:
```diff
@@ -0,0 +1,217 @@
+		return 'twa';
+	}
+
+	// Check if running in standalone mode
```
Comments like this one do not add additional value. The code is speaking enough. Remove this comment.

thlehmann-ionos commented on src/lib/IONOS/services/pwa.ts:
```diff
@@ -0,0 +1,217 @@
+ * Detects the current PWA display mode
+ */
+export function getPWADisplayMode(): PWADisplayMode {
+	// Check if running as TWA (Trusted Web Activity)
```
Note: This is a partly useful comment as it mentions the keyword "TWA", which is helpful when reading documentations.

thlehmann-ionos commented on src/lib/IONOS/services/pwa.ts:
```diff
@@ -0,0 +1,217 @@
+/**
+ * Tracks user engagement for PWA install prompt timing
+ */
+export function trackUserEngagement(): void {
```
This seems to be only relevant for the test/playground page/component. This should not be included in the service. Either remove altogether or refactor it out to the demo page/component.

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationBanner.svelte:
```diff
@@ -1,22 +1,51 @@
+			// For browsers that support beforeinstallprompt
+			deferredPrompt.prompt();
+			deferredPrompt.userChoice.then((choiceResult: any) => {
+				if (choiceResult.outcome === 'accepted') {
```
The accepted case is never handled and it's always dismissed.

thlehmann-ionos commented on src/lib/IONOS/components/common/Dialog.svelte:
```diff
@@ -41,6 +48,25 @@
+		on:close={() => {
+			show = false;
+		}}
+		on:transitionend={onTransitionEnd}
```
Triggering this event here too is probably redundant and it would be triggered twice (by both dialogs).

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -50,16 +66,108 @@
+		);
+
+		// For iOS devices, show prompt after a delay if not dismissed
+		if (isIOSDevice() && isSafari() && shouldShowPWAPrompt()) {
```
Android is not handled, why?

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -50,16 +66,108 @@
+				if (shouldShowPWAPrompt()) {
+					addPWANotification();
+				}
+			}, 3000); // Show after 3 seconds
```
This is a zero-value comment. Why is it deferred though?

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -50,16 +66,108 @@
+		<NotificationBanner
+			{notification}
+			{deferredPrompt}
+			on:dismiss={dismissHandler}
```
From `NotificationBanner` perspective these events are confusing.

thlehmann-ionos commented on src/lib/IONOS/services/pwa.ts:
```diff
@@ -0,0 +1,217 @@
+	}
+
+	// For iOS Safari, always show if conditions are met
+	if (isIOSDevice() && isSafari()) {
```
This check seems redundant as the conditions was already checked and handled by `isPWAInstallSupported()` one statement above.

thlehmann-ionos commented on src/lib/IONOS/services/pwa.ts:
```diff
@@ -0,0 +1,217 @@
+/**
+ * Checks if we should show the PWA install prompt
+ */
+export function shouldShowPWAPrompt(deferredPrompt: BeforeInstallPromptEvent | null = null): boolean {
```
The fact that it can be called with an optional event leads to confusion in some places where this function is called as at the calling point it does not work as expected.

thlehmann-ionos commented on src/lib/IONOS/services/pwa.ts:
```diff
@@ -0,0 +1,217 @@
+
+export interface BeforeInstallPromptEvent extends Event {
+	readonly platforms: string[];
+	readonly userChoice: Promise<{
```
This should be its own type. In one place this property was cast to `any`. Its type could and should be declared in that place (which also gives a linter warning on `any`).

thlehmann-ionos reviewed (changes requested):
Let's discuss.

thlehmann-ionos committed:
feat(Dialog): add outside-of-dialog close button

* Added external close-X below the content element
* Removed overflow-hidden to allow new close-X to leave the container

Notes:

* Works for the privacy links banner
* It should have a prop to enable this explicitly
* Maybe it's better to have a special dialog fo such use cases
* Maybe it should hide the inner button on mobile breakpoint and show
  the other one instead or maybe the inner one can be repurposed just
  by conditions

Refs: PRODAI-380

thlehmann-ionos committed:
refac(Dialog): refactor close action to named function

To allow reuse in another place.

Refs: PRODAI-380

bromiesTM committed:
c/chore(translations): add missing translation key

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

bromiesTM committed:
c/fix(pwa): dont dismiss on dialog close

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

bromiesTM committed:
c/fix(PwaDialog): fix external close button location & backdrop

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

bromiesTM committed:
c/refactor(pwa): remove comments without value

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

bromiesTM committed:
style(styleguide): make mobile compatible

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

bromiesTM committed:
c/feat: move pwa deferredPrompt to store

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

bromiesTM committed:
style(pwa dialog): improve spacing

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

bromiesTM committed:
refactor(pwa): handle pwa install in manager

Signed-off-by: Kai Henseler <kai.henseler@strato.de>
